src/commonMain/kotlin/arrow.exact/ExactDsl.kt

```diff
@@ -0,0 +1,30 @@
+package arrow.exact
+
+import arrow.core.*
+
+internal class AndExact<A, B, C>(
+  private val exact1: Exact<A, B>,
+  private val exact2: Exact<B, C>
+) : Exact<A, C> {
+
+  override fun fromOrEither(value: A): Either<ExactError, C> {
+    return exact1.fromOrEither(value)
+      .flatMap { exact2.fromOrEither(it) }
+  }
+}
+
+fun <A, B> exact(predicate: Predicate<A>, constructor: (A) -> B): Exact<A, B> {
```

Comment:
@ILIYANGERMANOV Unfortunately we can't implement it in such a way, kotlin compiler thinks that after `{` there must be a class body, not a lambda, so eventually there is no difference in arguments order. On the other hand `exact` can be used not only in delegation but also as regular method call:

``` kotlin
fun main() {
  exact(::NotBlankString) {
    it.isNotBlank()
  }
}
```

But it is rather synthetic example, doesn't look like a real use case

@nomisRev I am not quite familiar with Raise api, can you tell me where to look for implementation examples?

Reply:
Thank you for the explanation @ustits! I did all the code examples from my Fold w/o an actual IDE/compiler - apologies for that! Regarding `Raise` - I'd recommend:
- https://arrow-kt.io/learn/typed-errors/working-with-typed-errors/
- https://github.com/arrow-kt/arrow/pull/3038 (I've put some KDoc of Raise with short examples if you're looking for a TL;DR;) 

**Raise TL;DR;**
- like `throws` but type-safe, in order to raise you need a `Raise<A>` context
- to "catch" raise use `recover`
- interoperable with `Either`
- @nomisRev can give you a better explanation and details but that should be enough to get you started

Reply:
Thanks @ILIYANGERMANOV , can you tell me, am I moving in the right direction? As I understood I need to use `recover` on the constraint and map the result onto `Either`. Btw in such scenario probably we don't need a constructor as an argument:

``` kotlin
fun <A, B> exact(constraint: Raise<ExactError>.(A) -> B): Exact<A, B> {
  return object : Exact<A, B> {
    override fun from(value: A): Either<ExactError, B> {
      return recover({
        constraint(value).right()
      }) {
        error -> error.left()
      }
    }
  }
}
```

Reply:
@ustits yes, it looks good to me! Thank you for making the change :)
A valid example usage would look like this, right?
```kotlin
companion object : Exact<String, NotBlankTrimmedString> by exact({
   ensureNotNull(it.trim().takeIf(String::isNotBlank)?.let(::NotBlankTrimmedString)) { 
       ExactError("Cannot be blank!")
   }
})
```

```kotlin
companion object : Exact<Float, Percent> by exact({
   Percent(it.coerceIn(0f, 1f))
})
```

I like this API a lot! It's a lot simpler when we get rid of the constructor.
The constructor can be required only for the predicate `exact` case.

Reply:
@ustits instead of using `recover` with `left` and `right` you can also just use the `either` block. It works through `Raise` 😉. TL;DR `Raise` is a higher-level abstraction that can work over errors of `E`. So for `Either` it's just `E`, but for `Option` it's for example `None` or `typealias Null = Nothing?` for _nullable types`. 

```kotlin
      return either { constraint(value) }
```